File: fm-dddrive/src/main/java/io/dddrive/core/enums/model/base/EnumConfigBase.java
package io.dddrive.core.enums.model.base;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

public abstract class EnumConfigBase {

	private static final AtomicInteger configCounter = new AtomicInteger(0);
	private static final Set<EnumerationBase<?>> enumerations = new HashSet<>();

	protected static boolean isInConfig() {
		return configCounter.get() > 0;
	}

	static void addEnum(EnumerationBase<?> enumeration) {
		enumerations.add(enumeration);
	}

	static void assignEnumItems() {
		for (EnumerationBase<?> enumeration : enumerations) {
			enumeration.assignItems();
		}
		enumerations.clear();
	}

	protected void startConfig() {
		configCounter.incrementAndGet();
	}

	protected void endConfig() {
		configCounter.getAndUpdate(counter -> {
			if (counter == 1) {
				assignEnumItems();
			}
			return counter - 1;
		});
	}

}

File: fm-dddrive/src/main/java/io/dddrive/enums/model/base/EnumConfigBase.kt
package io.dddrive.enums.model.base

import java.util.concurrent.atomic.AtomicInteger
import java.util.function.IntUnaryOperator

abstract class EnumConfigBase {

	protected fun startConfig() {
		configCounter.incrementAndGet()
	}

	protected fun endConfig() {
		configCounter.getAndUpdate(
			IntUnaryOperator { counter: Int ->
				if (counter == 1) {
					assignEnumItems()
				}
				counter - 1
			},
		)
	}

	companion object {

		private val configCounter = AtomicInteger(0)
		private val enumerations: MutableSet<EnumerationBase<*>> = HashSet<EnumerationBase<*>>()

		fun isInConfig(): Boolean = configCounter.get() > 0

		fun addEnum(enumeration: EnumerationBase<*>?) {
			enumerations.add(enumeration!!)
		}

		fun assignEnumItems() {
			for (enumeration in enumerations) {
				enumeration.assignItems()
			}
			enumerations.clear()
		}
	}

}
